compiler/src/dotty/tools/dotc/transform/Splicer.scala

```diff
@@ -64,15 +64,82 @@ object Splicer {
     */
   def checkValidMacroBody(tree: Tree)(implicit ctx: Context): Unit = tree match {
     case Quoted(_) => // ok
-    case _ => (new CheckValidMacroBody).apply(tree)
+    case _ =>
+      def checkValidStat(tree: Tree): Unit = tree match {
+        case tree: ValDef if tree.symbol.is(Synthetic) =>
+          // Check val from `foo(j = x, i = y)` which it is expanded to
+          // `val j$1 = x; val i$1 = y; foo(i = y, j = x)`
+          checkIfValidArgument(tree.rhs)
+        case _ =>
+          ctx.error("Macro should not have statements", tree.sourcePos)
+      }
+      def checkIfValidArgument(tree: Tree): Unit = tree match {
+        case Block(Nil, expr) => checkIfValidArgument(expr)
+        case Typed(expr, _) => checkIfValidArgument(expr)
+
+        case Apply(TypeApply(fn, _), quoted :: Nil) if fn.symbol == defn.InternalQuoted_exprQuote =>
+          // OK
+
+        case TypeApply(fn, quoted :: Nil) if fn.symbol == defn.InternalQuoted_typeQuote =>
+          // OK
+
+        case Literal(Constant(value)) =>
+          // OK
+
+        case _ if tree.symbol == defn.QuoteContext_macroContext =>
+          // OK
+
+        case Call(fn, args)
+            if (fn.symbol.isConstructor && fn.symbol.owner.owner.is(Package)) ||
+               fn.symbol.is(Module) || fn.symbol.isStatic ||
+               (fn.qualifier.symbol.is(Module) && fn.qualifier.symbol.isStatic) =>
+          args.foreach(_.foreach(checkIfValidArgument))
+
+        case NamedArg(_, arg) =>
+          checkIfValidArgument(arg)
+
+        case SeqLiteral(elems, _) =>
+          elems.foreach(checkIfValidArgument)
+
+        case tree: Ident if tree.symbol.is(Inline) || tree.symbol.is(Synthetic) =>
+          // OK
+
+        case _ =>
+          ctx.error(
+            """Malformed macro parameter
+              |
+              |Parameters may be:
+              | * Quoted parameters or fields
+              | * References to inline parameters
+              | * Literal values of primitive types
+              |""".stripMargin, tree.sourcePos)
+      }
+      def checkIfValidStaticCall(tree: Tree): Unit = tree match {
+        case Block(stats, expr) =>
+          stats.foreach(checkValidStat)
+          checkIfValidStaticCall(expr)
+
+        case Typed(expr, _) => checkIfValidStaticCall(expr)
+        case Call(fn, args)
+            if (fn.symbol.isConstructor && fn.symbol.owner.owner.is(Package)) ||
+               fn.symbol.is(Module) || fn.symbol.isStatic ||
+               (fn.qualifier.symbol.is(Module) && fn.qualifier.symbol.isStatic) =>
+          args.flatten.foreach(checkIfValidArgument)
```

Comment:
Agreed the code here will diverge from `Call(..)` above, so there's no need to dedup now.